### PR TITLE
Address the outstanding review comments from #526

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch addresses the review comments left outstanding when
[#526](https://github.com/leynos/netsuke/pull/526) was closed with its
work landed on `main` (as `tests/makefile_test_target/rustflags.rs`).
Each comment was verified against the current implementation; three
still applied and are fixed here, and one is stale.

Fixed:

- **Shared `POLONIUS_FLAGS` resolution.** The
  `make_variable(…, "POLONIUS_FLAGS")` lookup was duplicated at three
  sites, and none rejected an empty value — every downstream assertion
  uses `contains`, which an empty string satisfies vacuously, so an
  empty definition would silently void the Polonius contract. A shared
  `polonius_flags` helper now rejects missing and empty definitions,
  and all three sites use it.
- **`assert_cmd::Command` for the `sh` spawn.** The expansion helper
  now uses `assert_cmd::Command` (an existing dev-dependency),
  preserving `env_remove("RUSTFLAGS")` and the conditional environment
  assignment for inherited values.
- **Non-Unix dead-code residue.** `Command`, `CALLER_RUSTFLAGS`,
  `DENY_WARNINGS`, and `expand` were already `#[cfg(unix)]`-gated on
  `main`, but the two policy fields are read only by the Unix
  behavioural tests, so a non-Unix test build would flag them dead
  under `-D warnings`; a `cfg_attr`-gated expectation closes that.

Skipped as stale:

- **Deriving rstest cases from `RUSTFLAGS_CASES` by index.** The
  current implementation has no rstest parametrization to
  deduplicate: the behavioural tests iterate the array directly, and
  `behavioural_every_rustflags_recipe_line_is_under_contract` forces
  every `RUSTFLAGS`-setting recipe line to carry a case, so adding an
  array entry extends coverage automatically — the outcome the
  comment sought.

## Review walkthrough

- The whole change is one file:
  [tests/makefile_test_target/rustflags.rs](https://github.com/leynos/netsuke/blob/4a8acbd/tests/makefile_test_target/rustflags.rs)
  — the `polonius_flags` helper and its three call sites, the
  `assert_cmd::Command` import in the `#[cfg(unix)]` block, and the
  `cfg_attr` expectations on the two `RustflagsCase` policy fields.

## Validation

- `make check-fmt`: pass
- `make lint` (Clippy + Whitaker, warnings denied): pass
- `make test` (full workspace nextest + doctests): pass, including
  the 14 `makefile_test_target` contract tests

## Summary by Sourcery

Consolidate and harden handling of POLONIUS_FLAGS in the makefile RUSTFLAGS contract tests while improving Unix/non-Unix test configuration hygiene.

Enhancements:
- Introduce a shared helper to resolve POLONIUS_FLAGS and enforce that it is present and non-empty before running assertions.
- Annotate RustflagsCase policy fields with cfg_attr-based dead_code expectations for non-Unix builds to reflect their Unix-only usage.

Tests:
- Refactor RUSTFLAGS behaviour tests to use the new POLONIUS_FLAGS helper consistently across all call sites.
- Switch Unix shell expansion in the tests to use assert_cmd::Command instead of std::process::Command while preserving the existing environment handling.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The new empty-value guard in `polonius_flags` has no direct test; all three callers use the real non-empty `Makefile`, so removing the guard would leave the suite passing. | Add unit tests for missing and empty `POLONIUS_FLAGS`, and assert the expected error messages; keep the existing behavioural tests for non-empty values. |
| Module-Level Documentation | ❌ Error | Expand the module docstring: `//! Contract model...` states the subject but not its test utility, operation, or relationship to the parent Makefile contract module. | Add module-level `//!` paragraphs explaining extraction and shell expansion of Makefile `RUSTFLAGS` recipes, contract assertions, and integration with `makefile_test_target`. |
| Testing (Unit And Behavioural) | ⚠️ Warning | `polonius_flags` adds missing and empty-value errors, but no test supplies either case; all three callers use the real, non-empty Makefile value. | Add unit tests for missing and empty or whitespace-only `POLONIUS_FLAGS`, and assert that each case returns an error. |
| Testing (Property / Proof) | ⚠️ Warning | `polonius_flags` introduces a non-empty-input invariant over arbitrary Makefile text, but the module has no proptest or Kani coverage for missing, empty, whitespace, and valid values. | Add a bounded `proptest` for `polonius_flags` that generates missing, whitespace-only, and non-empty `POLONIUS_FLAGS` definitions and asserts the acceptance invariant. |

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat o11y, code safety, documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 2 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The new empty-value guard in `polonius_flags` has no direct test; all three callers use the real non-empty `Makefile`, so removing the guard would leave the suite passing. | Add unit tests for missing and empty `POLONIUS_FLAGS`, and assert the expected error messages; keep the existing behavioural tests for non-empty values. |
| Module-Level Documentation | ❌ Error | Expand the module docstring: `//! Contract model...` states the subject but not its test utility, operation, or relationship to the parent Makefile contract module. | Add module-level `//!` paragraphs explaining extraction and shell expansion of Makefile `RUSTFLAGS` recipes, contract assertions, and integration with `makefile_test_target`. |
| Testing (Unit And Behavioural) | ⚠️ Warning | `polonius_flags` adds missing and empty-value errors, but no test supplies either case; all three callers use the real, non-empty Makefile value. | Add unit tests for missing and empty or whitespace-only `POLONIUS_FLAGS`, and assert that each case returns an error. |
| Testing (Property / Proof) | ⚠️ Warning | `polonius_flags` introduces a non-empty-input invariant over arbitrary Makefile text, but the module has no proptest or Kani coverage for missing, empty, whitespace, and valid values. | Add a bounded `proptest` for `polonius_flags` that generates missing, whitespace-only, and non-empty `POLONIUS_FLAGS` definitions and asserts the acceptance invariant. |

**Comment by leynos** [comment]
@coderabbitai resume

File: tests/makefile_test_target/rustflags_polonius_tests.rs [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+}
+
+#[rstest]
+fn polonius_flags_names_the_empty_value() {
+    let error = polonius_flags("POLONIUS_FLAGS ?=  \n").expect_err("empty definition should fail");
+    assert!(
+        error.to_string().contains("should not be empty"),
+        "empty definition should report the expected message; got {error:?}"
+    );
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/makefile_test_target/rustflags_polonius_tests.rs:57]: @coderabbitai Agreed and done in 2964a3a. `polonius_flags_names_the_missing_variable` and `polonius_flags_names_the_empty_value` are now a single `#[rstest]`, `polonius_flags_names_the_rejection_cause`, with `#[case::missing_variable("A ?= b\n", "should define POLONIUS_FLAGS")]` and `#[case::empty_value("POLONIUS_FLAGS ?=  \n", "should not be empty")]`. Both the input and the expected diagnostic substring come from the case parameters, and the assertion message quotes the makefile text so a failure still names the offending input. Both cases pass under `cargo nextest run --test makefile_test_target -E 'test(polonius)'`.

File: tests/makefile_test_target/rustflags_polonius_tests.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+/// Strategy for the `POLONIUS_FLAGS` value slot: absent, whitespace-only, or
+/// a non-empty flag-like token.
+fn arb_polonius_value() -> impl Strategy<Value = Option<String>> {
+    prop_oneof![
+        Just(None),
+        "[ \t]{0,4}".prop_map(Some),
+        "[-A-Za-z0-9=+.]{1,24}".prop_map(Some),
+    ]
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/makefile_test_target/rustflags_polonius_tests.rs:68]: @coderabbitai Valid finding, fixed in 2964a3a. The non-empty branch of `arb_polonius_value` is now a tuple strategy over optional leading whitespace, the flag token, and optional trailing whitespace, mapped to `Some(format!("{prefix}{flag}{suffix}"))`, leaving the absent and whitespace-only branches untouched.

The property already compares against `value.as_deref().map(str::trim)`, so padded generations now genuinely exercise the trimming contract: an untrimmed return from `polonius_flags` fails `prop_assert_eq!(resolved, trimmed, ...)` rather than passing vacuously. The doc comment on the strategy records why the padding is there. Verified green with a focused `cargo nextest run --test makefile_test_target` filtered to the Polonius tests.

File: tests/makefile_test_target/rustflags.rs [hunk 3]
```diff
@@ -37,8 +52,19 @@ struct RustflagsCase {
...
+    #[cfg_attr(
+        not(unix),
+        expect(dead_code, reason = "read only by Unix expansion tests")
+    )]
     warning_policy: WarningPolicy,
     /// How the recipe handles a caller-supplied value.
+    #[cfg_attr(
+        not(unix),
+        expect(dead_code, reason = "read only by Unix expansion tests")
+    )]
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/makefile_test_target/rustflags.rs:67]: @coderabbitai Respectfully declining this one; the current form is the smaller and better-tracked of the two options.

On "untracked": `#[expect(dead_code, reason = "...")]` is the *tracked* attribute. Unlike `allow`, an unfulfilled `expect` fires `unfulfilled_lint_expectation`, so if the fields ever become live on non-Unix the attribute itself becomes a warning and gets removed. It is also `#[cfg_attr(not(unix), ...)]`, so on Unix — where every gate in this repository actually runs — no lint attribute exists at all. The current cost is two attributes carrying a written reason.

On the proposed alternative: `warning_policy` and `inheritance_policy` are initialised in nine `const fn` constructors on `RustflagsCase` plus one inline struct literal in `unit_rejects_escaped_shell_command_substitution` (which is deliberately not `#[cfg(unix)]`, since `shell_expression` never invokes a shell). Gating the fields therefore means `#[cfg(unix)]` on both enums, both field declarations, and all twenty initialiser expressions — twenty-four attributes replacing two, spread across the constructor table that is the file's readable centrepiece. That is a net loss in maintainability, which is the concern the finding is raised under.

There is also a correctness hazard in the swap: a `#[cfg(unix)]` field initialiser is easy to omit when a tenth case is added, and the omission is invisible on Unix until someone builds the test target for Windows. The `expect` form has no such failure mode — the fields are unconditionally present and unconditionally set.

Worth noting that no CI job compiles this test target for a non-Unix host today (the Windows jobs build release artefacts, not tests), so the attribute is defensive rather than load-bearing. That argues for keeping it minimal, not for expanding it. If the policy metadata ever grows a non-Unix reader, the attributes disappear on their own via the unfulfilled-expectation warning.

**Comment by leynos** [comment]
@coderabbitai Pre-merge checks reconciled against the current tree (head `8e75430`), row by row.

**Developer Documentation — ⚠️ Warning — accepted and actioned in `8e75430`.**

The finding was correct: `docs/developers-guide.md` documented the `RUSTFLAGS` recipe shape and `tests/support/makefile.rs`, but said nothing about the contract-test module that consumes them. A new "The Makefile `RUSTFLAGS` contract tests" subsection now sits alongside "Shared Makefile contract helpers" and covers the three things the resolution asked for:

- *The contract-test boundary.* The module layout — `tests/makefile_test_target.rs` as the crate root, with `rustflags.rs` and `rustflags_polonius_tests.rs` as children — and what the tests assert on. They expand each assignment with `printf '%s'` under `sh` and assert on the resulting flags rather than on recipe text, and the command a recipe would run is never executed, so nothing here invokes Cargo, Kani, nextest, or Dylint. It also records why the behavioural tests and the policy fields they read are gated on `#[cfg(unix)]` (expansion needs a shell) while the parsing tests are not, and names the two guards that keep the model honest: rejection of unresolved Make variables and shell command substitution, and the completeness test that fails when a recipe sets `RUSTFLAGS` without a matching case.
- *`polonius_flags` rejection rules.* Both states are documented with their diagnostics — missing definition ("Makefile should define POLONIUS_FLAGS") and empty-after-trimming ("POLONIUS_FLAGS should not be empty") — together with the reason they matter: every assertion built on the resolved value uses `contains`, which an empty needle satisfies vacuously, so an empty value would silently void the Polonius contract rather than fail it. The success contract (the trimmed value) and the bounded proptest pinning that acceptance invariant are recorded too.
- *Sibling-module integration.* Why the children are not separate Cargo test targets, how they reach the shared helpers via `use super::{read_repo_file, target_recipe}`, and the reuse policy: general parsing helpers graduate to `tests/support/makefile.rs` once a second contract test needs them, whereas model types such as `RustflagsCase` stay private to the contract they describe.

`markdownlint-cli2` reports zero errors on the file.

**Architectural Complexity And Maintainability — ❓ Inconclusive — validated locally; no failure found.**

Performing the validation myself, scoped to this pull request's three files:

- *CodeScene delta.* `cs delta main` against the current working tree reports **"No issues found!"** — no new code health issues introduced by the branch. CI's CodeScene Code Health Review job agrees (passing).
- *The helper and its call sites.* `polonius_flags` is a nine-line `pub(crate)` function: one `make_variable` lookup, one `ensure!`, one `Ok`. It has four call sites — `shell_expression` plus the two `#[cfg(unix)]` behavioural tests in `rustflags.rs`, and the guard tests in `rustflags_polonius_tests.rs`. Promoting it from private to `pub(crate)` is what makes the sibling module able to test the guard directly, which is the point of the change.
- *Existing utilities.* No duplication was introduced. The generic parsing (`repo_root`, `read_repo_file`, `parse_rule`, `target_prerequisites`, `target_recipe`) stays in `tests/support/makefile.rs`; `polonius_flags` and `make_variable` are specific to this contract and correctly stay with it. Duplication *was removed* this round: two near-identical diagnostic tests collapsed into one parameterised `#[rstest]` in `2964a3a`.
- *Test-module integration.* One `mod rustflags_polonius_tests;` declaration in the crate root and one `use super::rustflags::polonius_flags;` in the child. No cyclic or ambient coupling; the child needs no Makefile access of its own because it drives the guard with synthetic text.

I could find nothing here that fails the check, so I am treating the row as passing on the evidence above rather than as outstanding work. If a specific complexity concern was behind the inconclusive verdict, name it and I will address it directly.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Validating the two inconclusive pre-merge rows against head 4d1053c, row by row:

**User-Facing Documentation** — not applicable to this branch. The full diff against main touches only developer-facing material: the Makefile `RUSTFLAGS` contract tests (`tests/makefile_test_target.rs`, `tests/makefile_test_target/rustflags.rs`, `tests/makefile_test_target/rustflags_polonius_tests.rs`), the developers' guide, the spelling configuration overlay, and two prose corrections in design/execplan documents. No user-visible behaviour, command, or installation flow changes, so there is no user-facing documentation to update.

**Developer Documentation** — covered. Commit 8e75430 added a 54-line subsection to `docs/developers-guide.md` documenting the contract-test boundary, the `polonius_flags` rejection rules (including why an empty value would void the contract vacuously), and the sibling-module integration and reuse policy for the test helpers this branch touches.

**Comment by buzzybee-df12** [comment]
@coderabbitai review
